Control > Explorer - use RemoveGenericItemModal + API for deletions

himdel opened the pull request:
This changes all deletions in Control > Explorer (alert_profiles, alerts, conditions, miq_actions, policies, policy_profiles)  to use the API.

Continuing @mzazrivec 's effort (https://github.com/ManageIQ/manageiq-ui-classic/pull/6812 and others), this uses `RemoveGenericItemModal` on the UI side.

Fixes https://github.ibm.com/IBMPrivateCloud/CP4MCM/issues/10722 + https://github.ibm.com/IBMPrivateCloud/CP4MCM/issues/10521 by not using the uuid in `name`, but preferring the name in `description` when asking for confirmation - added a `display_field` option for `RemoveGenericItemModal` to achieve that.

himdel commented:
**alert_profiles**

TODO


**alerts**

TODO


**conditions**

![condition](https://user-images.githubusercontent.com/289743/87698969-b89a8900-c783-11ea-9002-3fa29b1e8856.png)
![condition2](https://user-images.githubusercontent.com/289743/87698975-b9cbb600-c783-11ea-89ca-ceecffae4c9a.png)


**miq_actions**

![action](https://user-images.githubusercontent.com/289743/87698937-b0424e00-c783-11ea-96f1-273b5c0aa7fe.png)
![action2](https://user-images.githubusercontent.com/289743/87698941-b1737b00-c783-11ea-95a4-89d176fc70f1.png)


**policies**

![policy](https://user-images.githubusercontent.com/289743/87698993-be906a00-c783-11ea-8a5d-0634358c5cbb.png)
![policy2](https://user-images.githubusercontent.com/289743/87698997-c05a2d80-c783-11ea-9d3e-faab7424be3b.png)


**policy_profiles**

![policy_profile](https://user-images.githubusercontent.com/289743/87699007-c3551e00-c783-11ea-8b2b-54cb48e7a52b.png)
![policy_profile2](https://user-images.githubusercontent.com/289743/87699018-c4864b00-c783-11ea-9e53-d32b041114d3.png)


(**events** are not deletable)

himdel commented:
@miq-bot remove_label wip

Looks like we have API for all miq_policy entities, it can already delete, so the only real change was allowing the modal to use `description` instead of `name` when needed.

mzazrivec commented:
When I delete a policy profile, the modal dialog nicely shows and deletes the profile, but the deleted profile won't disappear from the tree rendered on the left, since the tree is not being re-rendered. We could use `redirect_url` here instead to trigger full-page reload, but not sure if other errors wouldn't show (missing record & stuff).

I think the other deletions will show similar problem.

himdel commented:
Ah, good point. It looks like there's already a `reload` endpoint, with

```
-          miqAjax(`tree_select?id=${treeSelect}`).then(() => miqFlashSaved());
+          miqAjax(`reload?id=${treeSelect}`)
+            .then(() => miqFlashSaved());
```

it works, but right now it also flashes..

![20200717175202](https://user-images.githubusercontent.com/289743/87816078-4181f500-c856-11ea-83be-8de910a6401c.png)

Looking.. :)

himdel commented:
@miq-bot add_label wip

himdel commented:
So, how this works..

`reload` only sets id to the deleted item and calls `tree_select`,
`valid_active_node` then triggers tree replacement (but only if the id no longer exists, using id=root won't trigger it),
and it also triggers the message, unless `@report_deleted` is set, which is done in 2 controllers (non-api deletions).

`valid_active_node` is also the only place in the app which sets `@replace_trees`, used by `replace_right_cell` to actually replace the right trees.

So, options:
* change the called method from `tree_select?id=root` to just `reload?deleted=true`, and set `@report_deleted = true` when `params[:deleted]` - that should work everywhere
* add an explicit `replace_trees` param, and have `tree_select` accept it and force the tree reload this way - but this tree_select is implemented in each controller, so this would only affect miq_policy

Trying to go with option 1 :)

himdel commented:
Seems to work...

![fix](https://user-images.githubusercontent.com/289743/88203747-d1021c00-cc39-11ea-8769-53c9aaa8c2ef.png)

himdel commented:
@miq-bot remove_label wip

mzazrivec commented:
The CI failure is related:
```
FAIL  app/javascript/spec/remove-generic-item-modal/remove-generic-item-modal.spec.js
  ● RemoveGenericItemModal › removeItems() works correctly
    fetch-mock: No fallback response defined for POST to /api/undefined/123
      10 |   };
      11 | 
    > 12 |   return fetch(options.url, fetchOpts)
         |          ^
      13 |     .then(responseAndError(options));
      14 | }
      15 | 
```

mzazrivec commented:
Removing condition from under a policy profile still uses the old removal mechanism. Is that intended?

![Screenshot from 2020-07-23 12-52-16](https://user-images.githubusercontent.com/6648365/88278818-52fa4f80-cce3-11ea-8e9a-5eb1f1f1a030.png)

himdel commented:
> Removing condition from under a policy profile still uses the old removal mechanism. Is that intended?

It is intentional, that's not actually a deletion, just breaking the db association, so we can't use the same API.

himdel commented:
> The CI failure is related

Thanks, hopefully now :)

himdel commented:
Test failure unrelated now, needs ManageIQ/manageiq-ui-classic#7233
